python/plugins/processing/algs/qgis/RasterLayerUniqueValuesCount.py

```diff
@@ -0,0 +1,119 @@
+    Copyright            : (C) 2017 by Mathieu Pellerin
+    Email                : nirvn dot asia at gmail dot com
+***************************************************************************
+*                                                                         *
+*   This program is free software; you can redistribute it and/or modify  *
+*   it under the terms of the GNU General Public License as published by  *
+*   the Free Software Foundation; either version 2 of the License, or     *
+*   (at your option) any later version.                                   *
+*                                                                         *
+***************************************************************************
+"""
+
+__author__ = 'Mathieu Pellerin'
+__date__ = 'October 2017'
+__copyright__ = '(C) 2017, Mathieu Pellerin'
+
+# This will get replaced with a git SHA1 when you do a git archive
+
+__revision__ = '$Format:%H$'
+
+import codecs
+
+from qgis.core import (QgsProcessingParameterRasterLayer,
+                       QgsProcessingParameterBand,
+                       QgsProcessingParameterFileDestination,
+                       QgsProcessingOutputHtml)
+from processing.algs.qgis.QgisAlgorithm import QgisAlgorithm
+
+
+class RasterLayerUniqueValuesCount(QgisAlgorithm):
+
+    INPUT = 'INPUT'
+    BAND = 'BAND'
+    OUTPUT_HTML_FILE = 'OUTPUT_HTML_FILE'
+
+    def group(self):
+        return self.tr('Raster analysis')
+
+    def __init__(self):
+        super().__init__()
+
+    def initAlgorithm(self, config=None):
+        self.addParameter(QgsProcessingParameterRasterLayer(self.INPUT,
+                                                            self.tr('Input layer')))
+        self.addParameter(QgsProcessingParameterBand(self.BAND,
+                                                     self.tr('Band number'),
+                                                     1,
+                                                     self.INPUT))
+        self.addParameter(QgsProcessingParameterFileDestination(self.OUTPUT_HTML_FILE, self.tr('Unique values count'), self.tr('HTML files (*.html)'), None, True))
+        self.addOutput(QgsProcessingOutputHtml(self.OUTPUT_HTML_FILE, self.tr('Unique values count')))
+
+    def name(self):
+        return 'rasterlayeruniquevaluescount'
+
+    def displayName(self):
+        return self.tr('Raster layer unique values count')
+
+    def processAlgorithm(self, parameters, context, feedback):
+        layer = self.parameterAsRasterLayer(parameters, self.INPUT, context)
+        band = self.parameterAsInt(parameters, self.BAND, context)
+        outputFile = self.parameterAsFileOutput(parameters, self.OUTPUT_HTML_FILE, context)
+
+        uniqueValues = {}
+        width = layer.width()
+        height = layer.height()
+        current = 0
+        total = 100.0 / (width * height) if width > 0 and height > 0 else 0
+
+        rasterBlock = layer.dataProvider().block(1, layer.extent(), width, height)
```

Comment:
You always read first band here. Should be  changed to `layer.dataProvider().block(band, layer.extent(), width, height)`

Also this will work only for relatively small rasters which can be loaded into memory. For large datasets algorithm will be unusable. Better to read raster in chunks

Reply:
@alexbruy , thanks re band number.

Regarding the chunks proposal, I've tried the algorithm with a global forest change geotiff (40,000 x 40,000 pixels) and things worked fine. Since it's my primary intent, I'm inclined to leave things as is. 

How would you chunk things up otherwise? I'm not seeing this being taken into account with the raster calculator for e.g.

Reply:
If it works fine for large rasters, I'm fine with leaving it as is